### PR TITLE
Nested drag

### PR DESCRIPTION
**Pull request by jweinstein92**
## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)

## PR Type
What kind of change does this PR introduce?

```
[x] Bugfix
[ ] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[ ] CI related changes
[ ] Documentation content changes
[ ] Other... Please describe:
```

## What is the current behavior?

Currently nested nodes (nodes having a parent) are unable to be rearranged. Similarly, nodes which have children are unable to be rearranged
Issue Number: N/A

## What is the new behavior?
Find nodes that are nested within other nodes when attempting to drag. If dragging a node containing other nodes, generate JSX for the children. Remove styling of `unlinkable` on dragging nodes. Allow callbacks for `dragNode` and `dragEdge`. Allow dragging a nested node outside of its parent.

## Does this PR introduce a breaking change?
```
[ ] Yes
[x] No
```

## Other information
Should I update the version number and the changelog? I presume that will be done once this is merged in and a release is made.

### CODE REVIEW

File: src/Canvas.tsx [hunk 1]
```diff
@@ -270,14 +276,74 @@ const InternalCanvas: FC<CanvasProps & { ref?: Ref<CanvasRef> }> = forwardRef(
...
+              }}
+              id={`${id}-node-${n.id}-node-drag`}
+            />
+          );
+        });
+      },
+      [animated, edge, id, node]
     );
 
+    const dragNodeData = useMemo(() => {
```
Comment (amcdnl) [inline src/Canvas.tsx:311]: Let's abstract this into a function out of this scope.

File: src/Canvas.tsx [hunk 2]
```diff
@@ -270,14 +276,74 @@ const InternalCanvas: FC<CanvasProps & { ref?: Ref<CanvasRef> }> = forwardRef(
       }
     }, [layout, xy]);
 
-    const dragNodeData = useMemo(
-      () =>
-        rest.dragNode
-          ? layout?.children?.find((c) => c.id === rest.dragNode.id)
-          : null,
-      [layout?.children, rest.dragNode]
+    const createDragNodeChildren = useCallback(
```
Comment (amcdnl) [inline src/Canvas.tsx:279]: Don't use useCallback to due perf thrashing

File: src/Canvas.tsx [hunk 3]
```diff
@@ -270,14 +276,74 @@ const InternalCanvas: FC<CanvasProps & { ref?: Ref<CanvasRef> }> = forwardRef(
...
+              animated={animated}
+              nodes={children}
+              childEdge={edge}
+              childNode={node}
+              {...n}
+              onDragStart={(event) => {
+                // @ts-ignore
+                setDragType(event.dragType);
+              }}
+              id={`${id}-node-${n.id}-node-drag`}
```
Comment (amcdnl) [inline src/Canvas.tsx:303]: id and key should match ( ideally )

File: src/Canvas.tsx [hunk 4]
```diff
@@ -270,14 +276,74 @@ const InternalCanvas: FC<CanvasProps & { ref?: Ref<CanvasRef> }> = forwardRef(
...
-      [layout?.children, rest.dragNode]
+    const createDragNodeChildren = useCallback(
+      (children: any) => {
+        if (!children || !Array.isArray(children)) {
+          return [];
+        }
+
+        return children.map(({ children, ...n }) => {
+          const element =
+            typeof node === 'function' ? node(n as NodeProps) : node;
```
Comment (amcdnl) [inline src/Canvas.tsx:287]: Would this be `node` or `dragNode`?

File: src/Canvas.tsx [hunk 5]
```diff
@@ -270,14 +276,74 @@ const InternalCanvas: FC<CanvasProps & { ref?: Ref<CanvasRef> }> = forwardRef(
...
+      const foundNode = findNestedNode(
+        rest.dragNode.id,
+        layout?.children,
+        parent
+      );
+      if (foundNode && foundNode.children) {
+        const nodeCopy = { ...foundNode };
+        // Node children is expecting a list of React Elements, need to create a list of elements
+        nodeCopy.children = createDragNodeChildren(nodeCopy.children);
+        // nodeCopy.children = nodeCopy.children.map(n => ({ ...n, id: `${n.id}-node-drag` }));
```
Comment (amcdnl) [inline src/Canvas.tsx:340]: dead code

File: src/Canvas.tsx [hunk 6]
```diff
@@ -270,14 +276,74 @@ const InternalCanvas: FC<CanvasProps & { ref?: Ref<CanvasRef> }> = forwardRef(
...
+        return nodeCopy;
+      }
+
+      return foundNode;
+    }, [createDragNodeChildren, layout?.children, rest.dragNode]);
```
Comment (amcdnl) [inline src/Canvas.tsx:345]: Why use `rest` and not `dragNode` ?

File: src/Canvas.tsx [hunk 7]
```diff
@@ -270,14 +276,74 @@ const InternalCanvas: FC<CanvasProps & { ref?: Ref<CanvasRef> }> = forwardRef(
...
+        }
+        return foundNode;
+      }
+
+      const foundNode = findNestedNode(
+        rest.dragNode.id,
+        layout?.children,
+        parent
+      );
+      if (foundNode && foundNode.children) {
```
Comment (amcdnl) [inline src/Canvas.tsx:336]:
```suggestion
      if (foundNode?.children) {
```

File: src/Canvas.tsx [hunk 8]
```diff
@@ -270,14 +276,74 @@ const InternalCanvas: FC<CanvasProps & { ref?: Ref<CanvasRef> }> = forwardRef(
...
+      if (!rest.dragNode) {
+        return null;
+      }
+
+      const { parent } = rest.dragNode;
+      if (!parent) {
+        const foundNode = layout?.children?.find(
+          (c) => c.id === rest.dragNode.id
+        );
+        if (foundNode && foundNode.children) {
```
Comment (amcdnl) [inline src/Canvas.tsx:321]:
```suggestion
        if (foundNode?.children) {
```

File: src/Canvas.tsx [hunk 9]
```diff
@@ -270,14 +276,74 @@ const InternalCanvas: FC<CanvasProps & { ref?: Ref<CanvasRef> }> = forwardRef(
...
 
+    const dragNodeData = useMemo(() => {
+      if (!rest.dragNode) {
+        return null;
+      }
+
+      const { parent } = rest.dragNode;
+      if (!parent) {
+        const foundNode = layout?.children?.find(
+          (c) => c.id === rest.dragNode.id
```
Comment (amcdnl) [inline src/Canvas.tsx:319]:
```suggestion
          n => n.id === rest.dragNode.id
```

File: src/Canvas.tsx [hunk 10]
```diff
@@ -270,14 +276,74 @@ const InternalCanvas: FC<CanvasProps & { ref?: Ref<CanvasRef> }> = forwardRef(
...
+        );
+        if (foundNode && foundNode.children) {
+          const nodeCopy = { ...foundNode };
+          // Node children is expecting a list of React Elements, need to create a list of elements
+          nodeCopy.children = createDragNodeChildren(nodeCopy.children);
+          // nodeCopy.children = nodeCopy.children.map(n => ({ ...n, id: `${n.id}-node-drag` }));
+          return nodeCopy;
+        }
+        return foundNode;
```
Comment (amcdnl) [inline src/Canvas.tsx:328]:
```suggestion

        return foundNode;
```

File: src/Canvas.tsx [hunk 11]
```diff
@@ -346,7 +412,7 @@ const InternalCanvas: FC<CanvasProps & { ref?: Ref<CanvasRef> }> = forwardRef(
                   childEdge={edge}
                   childNode={node}
                   {...n}
-                  onDragStart={event => {
+                  onDragStart={(event) => {
```
Comment (amcdnl) [inline src/Canvas.tsx:415]:
```suggestion
                  onDragStart={event => {
```

File: src/symbols/Node/Node.tsx [hunk 12]
```diff
@@ -151,8 +152,10 @@ export const Node: FC<Partial<NodeProps>> = ({
   const newX = x + offsetX;
   const newY = y + offsetY;
   const isLinkable = checkNodeLinkable(properties, enteredNode, canLinkNode);
-  const isMultiPort = dragType === 'multiportOnly' &&
+  const isMultiPort =
+    dragType === 'multiportOnly' &&
     ports?.filter((p) => !p.properties?.hidden).length > 1;
+  const isNodeDrag = useMemo(() => id.includes('node-drag'), [id]);
```
Comment (amcdnl) [inline src/symbols/Node/Node.tsx:158]: Overkill optimization

File: src/utils/helpers.ts [hunk 13]
```diff
@@ -57,3 +57,42 @@ export function getCoords({
...
+
+  if (parentId) {
+    const parentNode = children.find((c) => c.id === parentId);
+    if (parentNode?.children) {
+      return findNestedNode(nodeId, parentNode.children, parentId);
+    }
+  }
+
+  // Check for nested children
+  const nodesWithChildren = children.filter((c) => c.children);
```
Comment (amcdnl) [inline src/utils/helpers.ts:87]:
```suggestion
  const nodesWithChildren = children.filter(c => c.children);
```

File: src/utils/helpers.ts [hunk 14]
```diff
@@ -57,3 +57,42 @@ export function getCoords({
...
+    return;
+  }
+
+  let foundNode = children.find((c) => c.id === nodeId);
+  if (foundNode) {
+    return foundNode;
+  }
+
+  if (parentId) {
+    const parentNode = children.find((c) => c.id === parentId);
```
Comment (amcdnl) [inline src/utils/helpers.ts:80]:
```suggestion
    const parentNode = children.find(n => n.id === parentId);
```

File: src/utils/helpers.ts [hunk 15]
```diff
@@ -57,3 +57,42 @@ export function getCoords({
...
+export function findNestedNode(
+  nodeId: string,
+  children: any[],
+  parentId?: string
+) {
+  if (!nodeId || !children) {
+    return;
+  }
+
+  let foundNode = children.find((c) => c.id === nodeId);
```
Comment (amcdnl) [inline src/utils/helpers.ts:74]:
```suggestion
  let foundNode = children.find(n => n.id === nodeId);
```

Comment (amcdnl) [reply src/utils/helpers.ts:74]: You don't really use the assignment - u could make this a const and cleanup code below. Over assignment out of scope can be confusing.

File: src/utils/helpers.ts [hunk 16]
```diff
@@ -57,3 +57,42 @@ export function getCoords({
...
   return new Matrix2D().translate(tx, ty).scale(zoom).inverse();
 }
+
+/**
+ * Given a nodeId to find, a list of nodes to check against, and an optional parentId of the node
+ * find the node from the list of nodes
+ */
+export function findNestedNode(
+  nodeId: string,
+  children: any[],
```
Comment (amcdnl) [inline src/utils/helpers.ts:67]: Missing type

**Review by amcdnl — CHANGES_REQUESTED** [review]

File: src/Canvas.tsx (hunk 6 above)
Comment (jweinstein92) [reply src/Canvas.tsx:345]: `rest.dragNode` uses `useCanvas` dragNode property which is `NodeData<any>` vs `dragNode` is a JSX element

Comment (amcdnl) [reply src/Canvas.tsx:345]: Got it - let's give it a named import then. I try to avoid rest references.

File: src/utils/helpers.ts (hunk 16 above)
Comment (jweinstein92) [reply src/utils/helpers.ts:67]: There is no declared type for this. `children` in this case is the `ElkLayout.children` which has a type of `any[]`

File: src/Canvas.tsx [hunk 17]
```diff
@@ -264,19 +272,55 @@ const InternalCanvas: FC<CanvasProps & { ref?: Ref<CanvasRef> }> = forwardRef(
...
+    };
+
+    useEffect(() => {
+      if (dragNodeData.children) {
+        const nodeCopy = { ...dragNodeData };
+        // Node children is expecting a list of React Elements, need to create a list of elements
+        nodeCopy.children = createDragNodeChildren(nodeCopy.children);
+        setDragNodeDataWithChildren(nodeCopy);
+      }
+    }, [createDragNodeChildren, dragNodeData]);
```
Comment (amcdnl) [inline src/Canvas.tsx:323]: You are gonna need to add `useCallback` back - the effect will always thrash now :(

**Review by amcdnl — CHANGES_REQUESTED** [review]

File: src/symbols/Node/Node.tsx [hunk 18]
```diff
@@ -4,6 +4,7 @@ import React, {
   ReactElement,
   ReactNode,
   useEffect,
+  useMemo,
```
Comment (amcdnl) [inline src/symbols/Node/Node.tsx:7]: I don't think this is used anymore

File: src/utils/helpers.ts [hunk 19]
```diff
@@ -57,3 +57,62 @@ export function getCoords({
...
+
+    if (foundChild) {
+      return foundChild;
+    }
+  }
+
+  return {};
+}
+
+export function getDragNodeData(
```
Comment (amcdnl) [inline src/utils/helpers.ts:100]: Missing comments

File: src/utils/helpers.ts [hunk 20]
```diff
@@ -57,3 +57,62 @@ export function getCoords({
...
+  dragNode: NodeData,
+  children: any[] = []
+): { [key: string]: any } {
+  if (!dragNode) {
+    return {};
+  }
+
+  const { parent } = dragNode;
+  if (!parent) {
+    const foundNode = children?.find(n => n.id === dragNode.id);
```
Comment (amcdnl) [inline src/utils/helpers.ts:110]:
```suggestion
    return (children?.find(n => n.id === dragNode.id) || {});
```

File: src/utils/helpers.ts [hunk 21]
```diff
@@ -57,3 +57,62 @@ export function getCoords({
...
+
+    return foundNode || {};
+  }
+
+  const foundNode = findNestedNode(dragNode.id, children, parent);
```
Comment (amcdnl) [inline src/utils/helpers.ts:115]:
```suggestion
return findNestedNode(dragNode.id, children, parent);
```

File: stories/Drag.stories.tsx [hunk 22]
```diff
@@ -393,6 +393,121 @@ export const NodePortRearranging = () => {
...
+        }}
+        onNodeLink={(_event, from, to) => {
+          const result = removeAndUpsertNodes(
+            nodes,
+            edges,
+            from
+          );
+
+          // Update parents
+          if ((from.parent || to.parent) && from.parent !== to.parent) {
```
Comment (amcdnl) [inline stories/Drag.stories.tsx:491]: Shouldn't this happen in node link check?

File: src/utils/useEdgeDrag.ts [hunk 23]
```diff
@@ -3,6 +3,7 @@ import { EdgeSections } from '../symbols/Edge';
 import { NodeData, PortData } from '../types';
 import { DragEvent, NodeDragEvents, Position } from './useNodeDrag';
 import { Point2D } from 'kld-affine';
+import { NodeDragType } from 'symbols/Node';
```
Comment (amcdnl) [inline src/utils/useEdgeDrag.ts:6]: Need the .. pathing on it 

```suggestion
import { NodeDragType } from '../symbols/Node';
```

File: stories/Drag.stories.tsx (hunk 22 above)
Comment (jweinstein92) [reply stories/Drag.stories.tsx:491]: This doesn't affect the validity of a node link occurring. This is determining whether an updated parent is needed based on the new link that is going to happen

**Review by amcdnl — APPROVED** [review]

**Pull request merged**